Implement issue #8196: Leave on read all notifications involving a blocked person

nat-sch opened the pull request:
Hi, hope everyone is doing well!

This PR would implement the feature requested in #8196.

I went with the proposed alternative of marking the notifications as read instead of deleting them, because I think that's more trackable in case the notifications contained content that might be of interest to keep, such as proof of harassment, and also because I feel deleting them would be more unexpected for the user.

I didn't use a cuke since I didn't wanna add such a slow and high level test for this comparatively low impact feature, but I can add it if we deem it necessary.

I'll be waiting to hear your thoughts! 😄

nat-sch committed:
Test for NotificationService#read_all_involving

nat-sch committed:
Implement NotificationService#read_all_involving

nat-sch committed:
Test that BlockService#block reads all notes of blocked person

nat-sch committed:
Make BlockService#block read all notes of blocked person

nat-sch committed:
Test that the Person model refetches the note list

nat-sch committed:
Make Person model refetch note list

nat-sch commented:
I see that the job `(3.2, postgresql, other)` failed. I can run the failing command as `RAILS_ENV=test bin/rake jasmine:ci` on my machine, with Ruby 3.2.6 and over Postgres 14.13, without it failing. I get `954 specs, 0 failures, 4 pending specs`, like in the sucessful Jasmine jobs. I don't know why it failed on the CI 😞

denschub commented:
The test failure is a random intermittent, it looks like. I just re-triggered the specific suite.

nat-sch commented:
Thank you!! It looks like it was, now all checks have passed! 🚀

skyschub reviewed (approved):
Okay, first: thanks for your PR, and welcome to this project. It's rare that the first time contribution of anyone a) passes all lints, b) passes all tests, c) even introduces new tests. Kudos. :)

This looks fine to me, especially the client side. I'll hold off on @SuperTux88 in case he has some comments about the Ruby side, but it looks good to me.

SuperTux88 commented on app/services/notification_service.rb:
```diff
@@ -56,6 +56,17 @@ def notify(object, recipient_user_ids)
+    Notification
+      .for(@user)
+      .joins(:notification_actors)
+      .where(notification_actors: {person: person})
+      .find_each do |note|
+        note.unread = false
+        note.save
+      end
```
I'm wondering if it would be worth also filtering only still unread notifications? Or instead of changing each entity separately, maybe use `update_all(unread: false)`? This would skip the vailidation of the entity (as it directly calls the database and skips the rails logic) and pronto/rubocop will be unhappy about it. But there is no validation based on the read-state, if it was valid before, it's also still valid after being marked as read, so I think in this case it should be fine to ignore this rule.

However, I think the current implementation is fine for now for the current intended use-case, where people probably don't have a lot of already read notifications for a person they want to block. And also they probably don't have a lot of notifications overall yet for the person they want to block (at least not as many as it would make a notable performance difference).

nat-sch replied:
Ahh, insightful, thank you! I went with filtering the already read notes in the `where`.

I don't like sending a query that updates a row with exactly the same data, I didn't realize that Rails' optimizations didn't reach that far. Also, thinking about the use case, I'm not sure, but think that it will probably happen sometimes that a user wants to block another user with whom they have have been interacting for a while.

About the validations, I like the idea of not skipping them right now, because I want to make it easy to extend them in the future (therefore reducing the amount of places where they can be skipped).

SuperTux88 replied:
> after more research, maybe Rails doesn't send a query if the record didn't change upon save, but it will run the validations.

I'm also not 100% sure if rails really sends out the update or what it maybe automatically skips, but it will at least run something, and for a loop with potentially a lot of entries, I at least wanted to think about.

> About the validations, I like the idea of not skipping them right now, because I want to make it easy to extend them in the future (therefore reducing the amount of places where they can be skipped).

Yeah, I also thought about if it could be a problem for future validations, but I couldn't think of any potential useful validation where a notification could be valid if it's unread, but then invalid if it's read. But I'm fine with leaving it with the loop and update each notification separately. The `update_all` was only an idea that would enable it to do it with just one query. But with the read notifications filtered out, there shouldn't be a lot of notifications left in most cases, so this is fine. :+1:

SuperTux88 reviewed (approved):
Thanks for the PR. This looks good to me. I had one comment that came to my mind while reading the code, but it's more of a note/question and I think it's fine for now as it is. :+1:

nat-sch committed:
In NotificationService#read_all_involving, filter for only unread notifications

SuperTux88 commented on app/services/notification_service.rb:
```diff
@@ -56,6 +56,20 @@ def notify(object, recipient_user_ids)
+      .for(@user)
+      .joins(:notification_actors)
+      .where(
+        notification_actors: {person: person},
```
While thinking about this a bit more, another potential problem came to my mind. If multiple people comment on the same post, we aggregate the notifications (that's why it can have multiple actors). So there could be a notification "Person A, Person B and Person C commented on your post", and I think with the current logic, if I block Person B, it would mark this notification as read, and I might miss the comments of Person A and C.

As a user, my expected behavior would be, that it only marks the notification as read, where the blocked user is the only actor.

However this maybe also is only an edge case, when I decide to block somebody who commented on a lot of posts where I also have a lot of notifications from other users I didn't read yet. So maybe it's fine to ignore that in the first version and improve later. But if you find an easy way to select only notifications where there is only one actor (the blocked person), that would be cool to add.

nat-sch replied:
I agree, thank you! Also thank you for the explanation on the purpose of storing notification actors!

I implemented only fetching notifications that have exclusively the blocked person as their actor. Let me know what you think!

nat-sch replied:
Hi @SuperTux88, I hope you are doing well. I wanted to kindly follow up on this when you have a moment. Let me know if there's anything else I can do to move it forward! Thank you for all your help so far, and thank you in advance!

nat-sch committed:
Read notes only involving blocked person

Change the test for this new spec, and make it manageably simpler at the cost of being less generic; implement the spec using a subquery

SuperTux88 committed:
Remove the need of using a sub-query and refactor it into a single query

SuperTux88 reviewed (approved):
Approved from my side, but since I also did some changes now, I requested another review from @denschub for my commit.

skyschub reviewed (approved):
No objection, looks good to me. Thank you so much, both! 🧡

SuperTux88 commented:
Merged as 5ac83952d686ce97042521d8be3d5d3c1f970c69 5fd1199d2035612ec4ce38a797b877ba64bfcca9 7586cda5e0951dbd19b3ade877ab0b37e8e82543 9f31159ffd45100299df620d7280f02cc9c7ec6f 1e7b844c2ea068c37745e7f0e1ad629a2a4a2a8c c661c9ea0d535b059f455d70a6269af8ea1b784f f44dd680958f88143120fa66c310f891fa3de704 59ff286c6408c055411781d0a01b3b2e60122407 and d5b03e0220b8a32dfd60231615a21367ae69e75f

And thanks @nat-sch for the PR and sorry it took so long to review and merge.
